create LSP23 LinkedContractsFactory (#227)
lukso-network/LIPs: LSPs/LSP-23-LinkedContractsFactory.md

author: skimaharvey
discussions-to: <URL>
status: Draft
type: LSP
created: 2023-08-16
requires: ERC1167
---

## Table of Content

- [Simple Summary](#simple-summary)
- [Abstract](#abstract)
- [Motivation](#motivation)
- [Specification](#specification)
  - [Methods](#methods)
    - [deployContracts](#deploycontracts)
    - [deployERC1167Proxies](#deployerc1167proxies)
- [Implementation](#implementation)
- [Interface Cheat Sheet](#interface-cheat-sheet)
- [Copyright](#copyright)

## Simple Summary

This standard describes a smart contract factory that can deploy linked contracts and have an optional post-deployment execution.

## Abstract

The `LSP23LinkedContractsFactory` allows users to deploy primary and secondary contracts that are linked together. It also supports deploying contracts as `ERC1167` proxies. The factory can execute post-deployment modules to perform additional logic after deploying the contracts.

## Motivation

The ability to deploy linked contracts and execute post-deployment modules simplifies the process of deploying complex contract systems. This standard aims to provide a unified interface for deploying linked contracts and executing post-deployment logic.

> In the context of multi-chain Ethereum smart contract development, there are situations where a contract, used by an another contract, needs to be deployed at the same addresses across multiple chains. This presents a complex challenge as it requires deterministic address generation for the contracts across different chains. The `LSP23` standard addresses this issue by generating the salt for contract deployment within the function, ensuring a fully decentralized deployment process. This approach not only guarantees consistent contract addresses across multiple chains but also prevents address squatting, enhancing the security and integrity of the deployment process.

Reviewer: `used by an another contract` -> `used by another contract`

> In the context of multi-chain Ethereum smart contract development, there are situations where a contract, used by an another contract, needs to be deployed at the same addresses across multiple chains. This presents a complex challenge as it requires deterministic address generation for the contracts across different chains. The `LSP23` standard addresses this issue by generating the salt for contract deployment within the function, ensuring a fully decentralized deployment process. This approach not only guarantees consistent contract addresses across multiple chains but also prevents address squatting, enhancing the security and integrity of the deployment process.

Reviewer: Just wanted to clarify. What "function" are you talking about? 
The part of the sentence is "by generating the salt for contract deployment within the function".

Author: thanks for the feedback. totally reviewed this part to avoid repetition

---
lip: 23
title: LSP23 Linked Contracts Factory
author: skimaharvey
discussions-to: <URL>
status: Draft
type: LSP
created: 2023-08-16
requires: ERC1167
---

## Table of Content

- [Simple Summary](#simple-summary)
- [Abstract](#abstract)
- [Motivation](#motivation)
- [Specification](#specification)
  - [Methods](#methods)
    - [deployContracts](#deploycontracts)
    - [deployERC1167Proxies](#deployerc1167proxies)
- [Implementation](#implementation)
- [Interface Cheat Sheet](#interface-cheat-sheet)
- [Copyright](#copyright)

## Simple Summary

> Deploying smart contracts that need to interact with each other can be complicated. This is especially true when these contracts need to know each other's details at the time they are created. The LSP23 standard simplifies this process by providing a unified way to deploy such interdependent contracts. It also allows for additional actions to be taken automatically after the contracts are deployed, making the whole process more streamlined and less error-prone.

Reviewer 2: small suggestions

```suggestion
Deploying smart contracts that need to interact with each other can be complicated. This is especially true when these contracts need to know each other's details at the time they are created. The LSP23 standard simplifies this process by providing a unified way to deploy such interdependent contracts. It also allows for additional actions to be taken automatically after the contracts have been deployed, streamlining the whole process and making it less error-prone.
```

## Abstract

> The LSP23 Linked Contracts Factory standard introduces a unified interface for deploying interdependent smart contracts, also known as linked contracts. These contracts often require each other's addresses at the time of deployment, creating a circular dependency. LSP23 addresses this by allowing for the deployment of a primary contract and one or more secondary contracts, linking them together. The standard also supports the deployment of contracts as `ERC1167` minimal proxies and enables optional post-deployment modules for executing additional logic. This serves to simplify and standardize the deployment of complex contract systems.

Reviewer 2: ```suggestion
The LSP23 Linked Contracts Factory standard introduces an unified interface for deploying interdependent smart contracts, also known as **linked contracts**. These contracts often require each other's addresses at the time of deployment, creating a circular dependency. LSP23 addresses this by allowing for the deployment of a primary contract and one or more secondary contracts, linking them together. The standard also supports the deployment of contracts as [ERC1167](https://www.google.com/search?q=ERC1167&oq=ERC1167&aqs=chrome..69i57.2294j0j1&sourceid=chrome&ie=UTF-8) minimal proxies and enables optional post-deployment modules for executing additional logic. This serves to simplify and standardize the deployment of complex contract systems.
```

## Motivation

> While there are various methods for deploying smart contracts in the current ecosystem, the challenge of deploying interdependent contracts in a streamlined and secure manner remains. Custom solutions often exist, but they can be cumbersome and prone to errors. LSP23 aims to standardize this process, reducing the complexity and potential for mistakes. By generating the salt for contract deployment within the function, the standard also ensures a fully decentralized deployment process. This not only guarantees consistent contract addresses across multiple chains but also serves as a preventive measure against address squatting, enhancing the overall security and integrity of the deployment process.

Reviewer 2: Should we specify CREATE2 here too?

```suggestion
While there are various methods for deploying smart contracts in the current ecosystem, the challenge of deploying interdependent contracts in a streamlined and secure manner remains. Custom solutions often exist, but they can be cumbersome and error prone. LSP23 aims to standardize this process, reducing the complexity and potential for mistakes. By generating the CREATE2 salt for contract deployment within the function, the standard also ensures a fully decentralized deployment process. This not only guarantees consistent contract addresses across multiple chains but also serves as a preventive measure against _"address squatting"_, enhancing the overall security and integrity of the deployment process.
```

## Specification

### Linked Contracts Factory Setup

#### Deployment

> Before the deployment of the `LSP23LinkedContractsFactory` on any network, people should make sure that the Nick Factory is deployed on the same network.

Reviewer 2: ```suggestion
Before the deployment of the `LSP23LinkedContractsFactory` on any network, people should make sure that the [Nick Factory](https://yamenmerhi.medium.com/nicks-method-ethereum-keyless-execution-168a6659479c) is deployed on the same network.
```

##### Nick Factory Deployment

> The Nick Factory should be located at this address `0x4e59b44847b379578588920ca78fbf26c0b4956c` on the network. If there is no code on this address, it means that the contract is not deployed yet.

Reviewer 2: ```suggestion
The Nick Factory contract should be located at this address `0x4e59b44847b379578588920ca78fbf26c0b4956c` on the network. If there is no code on this address, it means that the contract is not deployed yet.
```

To deploy, the following raw transaction should be broadcasted to the network `0xf8a58085174876e800830186a08080b853604580600e600039806000f350fe7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf31ba02222222222222222222222222222222222222222222222222222222222222222a02222222222222222222222222222222222222222222222222222222222222222` after funding the deployer address: `0x3fab184622dc19b6109349b94811493bf2a45362` with `gasPrice (100 gwei) * gasLimit (100000)`.

Check [Nick's Factory repository](https://github.com/Arachnid/deterministic-deployment-proxy/tree/master) for more information.

##### LSP23LinkedContractsFactory Deployment

> After the deployment of Nick Factory on the network, the `LSP23LinkedContractsFactory` can be reproduced at the standardized address given sending the same salt and bytecode.

Reviewer 2: ```suggestion
After the deployment of the Nick Factory contract on the network, the `LSP23LinkedContractsFactory` can be reproduced at the standardized address given sending the same salt and bytecode.
```

> In order to create the UniversalFactory contract, one should send a transaction to the [Nick Factory] address with data field equal to [salt](#standardized-salt) + [bytecode](#standardized-bytecode).

Reviewer 2: Is this a mistake? Or this intended?

```suggestion
In order to create the `LSP23LinkedContractsFactory` contract, one should send a transaction to the [Nick Factory] address with data field equal to [salt](#standardized-salt) + [bytecode](#standardized-bytecode).
```

The address produced should be equal to `0x2300000A84D25dF63081feAa37ba6b62C4c89a30`.

> #### UniversalFactory Configuration

Reviewer 2: Same shouldn't the title be different to be `LSP23LinkedContractsFactory` ✅ ? Not `UniversalFactory` ❌ ?
